### PR TITLE
#7111 cleaned up one case of redundant CRC32 digest and byte[] creation

### PR DESCRIPTION
**Pull request by original-brownbear**
One step for #7111, we can't do all of this in one PR, there's many many spots :)

* Moved to reused `CRC32` instance
* Serialize into `DirectByteBuffer` and adjusted file writing accordingly
* Adjusted benchmark to show potential gain/loss but no effect thanks to file creation being so expensive that it outweighs the other operations

### CODE REVIEW

**Comment by original-brownbear** [comment]
@jordansissel pong :)

**Comment by original-brownbear** [comment]
@jordansissel unpong, get well soon this one is not urgent :)

File: logstash-core/benchmarks/src/main/java/org/logstash/benchmark/QueueBenchmark.java [hunk 1]
```diff
@@ -87,8 +87,8 @@ private static Settings settings() {
             .capacity(256 * 1024 * 1024)
             .queueMaxBytes(Long.MAX_VALUE)
             .elementIOFactory(MmapPageIO::new)
-            .checkpointMaxWrites(50_000)
-            .checkpointMaxAcks(50_000)
+            .checkpointMaxWrites(1024)
```
Comment (jordansissel) [inline logstash-core/benchmarks/src/main/java/org/logstash/benchmark/QueueBenchmark.java]: It's not clear why this change was made. Can you omit it or include the reason in the commit message?

File: logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java [hunk 2]
```diff
@@ -115,14 +116,18 @@ private Checkpoint read(BufferedChecksumStreamInput crcsi) throws IOException {
...
+        buf.clear();
+        buf.putShort((short)Checkpoint.VERSION);
+        buf.putInt(checkpoint.getPageNum());
+        buf.putInt(checkpoint.getFirstUnackedPageNum());
+        buf.putLong(checkpoint.getFirstUnackedSeqNum());
+        buf.putLong(checkpoint.getMinSeqNum());
+        buf.putInt(checkpoint.getElementCount());
+        buf.flip();
+        crc32.update(buf);
+        buf.position(BUFFER_SIZE - Integer.BYTES).limit(BUFFER_SIZE);
```
Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java]: the bytebuffer api is such a strange thing.

I'd kind of prefer using `buf.mark()` (before you call `flip`) then `buf.reset()` instead of `buf.position(...)` to avoid any math problems, but I leave it up to you to decide. (keep the `.limit` call)

File: logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java [hunk 3]
```diff
@@ -64,10 +63,12 @@ public Checkpoint write(String fileName, int pageNum, int firstUnackedPageNum, l
 
     @Override
     public void write(String fileName, Checkpoint checkpoint) throws IOException {
-        Path path = Paths.get(dirPath, fileName);
-        final byte[] buffer = new byte[BUFFER_SIZE];
         write(checkpoint, buffer);
-        Files.write(path, buffer, WRITE_OPTIONS);
+        buffer.flip();
+        try (FileOutputStream out = new FileOutputStream(Paths.get(dirPath, fileName).toFile())) {
```
Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java]: This also changes the file writing strategy from Files.write to FileOutputStream, can you tell me about this change?

File: logstash-core/benchmarks/src/main/java/org/logstash/benchmark/QueueBenchmark.java (hunk 1 above)
Comment (original-brownbear) [inline logstash-core/benchmarks/src/main/java/org/logstash/benchmark/QueueBenchmark.java]: @jordansissel see last point in the PR description, I added this to verify that this change does not come with unexpected performance issues, so I moved the ack interval to the default here. Makes sense right?

File: logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java (hunk 3 above)
Comment (original-brownbear) [inline logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java]: @jordansissel If I use `Files.write()` I have to operate in the on heap `byte[]` paradigm which forces another memory copy on me. This way I can go straight for writing out the buffer and retain the resilience by calling `sync`.

File: logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java (hunk 2 above)
Comment (original-brownbear) [inline logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java]: @jordansissel I think Java will disappoint you here :(

```java
    public final Buffer flip() {
        limit = position;
        position = 0;
        mark = -1;
        return this;
    }
```

marks aren't retained in `flip` so it won't help us here right?

File: logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java (hunk 3 above)
Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java]: I had a curiosity, so I looked at the FileOutputStream.getChannel().write() effect. It calls sun.nio.ch.IOUtil.write() which will create a ByteBuffer if the given buffer is not a DirectBuffer. Cool -- this looks like it indeed avoids creating extra byte array.

Given this constraint, I think we need a comment on the `allocateDirect` call (line 34) to indicate this necessary in order to achieve your zero-copy goal. A comment something like (on line 34) `// We use a DirectBuffer here in order to achieve a zero-copy checkpoint write`.

Additionally for safety (regarding future maintenance), add an `assert buffer.isDirect()` at the top of the `write` method?

I am OK with your proposed change to use FileOutputStream, and my above comments are aimed at helping us not accidentally break your zero-copy intent.

File: logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java (hunk 2 above)
Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java]: Buffer api continues to be a strange and mysterious thing. Good idea to check that the mark is destroyed on flip.

File: logstash-core/benchmarks/src/main/java/org/logstash/benchmark/QueueBenchmark.java (hunk 1 above)
Comment (original-brownbear) [inline logstash-core/benchmarks/src/main/java/org/logstash/benchmark/QueueBenchmark.java]: @jordansissel sure sec :)

File: logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java (hunk 3 above)
Comment (original-brownbear) [inline logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java]: @jordansissel I like putting a comment here explaining the use of a direct buffer because of I/O (and actually also the CRC32 call ... this is a lot less obvious to only be 0GC using a direct buffer imo :)).
I don't really like `assert` tbh, it's such a standard spot in Java 8 (Direct way caught up to Heap performance wise, even for serializing tiny old `int` the direct one is ahead on OSX for me). 
Seems like this invites a bunch of weird (in the sense that I have never seen anyone asserting `isDirect() == true` anywhere) `assert` all over the place. I think this is really something that should be looked out for in code reviews instead (which would be made easy by having the comment I guess :D).

Comment (original-brownbear) [inline logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java]: Just a comment ok ? :)

**Comment by original-brownbear** [comment]
@jordansissel comment added as requested and commit message fixed to include the benchmark change :)

File: logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java (hunk 3 above)
Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java]: My thought on `assert` is that we are relying on a property of an object that cannot be checked by the compiler, which kind of stinks. 

Having an `assert` makes our tests fail if we miss something in code review (we miss things frequently, and that's ok, we are human). Similar to why I like tests because we cannot catch everything in code review (imperfect humans with finite time and attention available).

I am OK without an assert given the comment is a good hint.
